bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/ColumnViewer.java

```diff
@@ -817,4 +840,208 @@ protected void setBusy(boolean busy) {
+		}
+
+		if (sorted.length <= itemsLimit) {
+			return sorted;
+		}
+
+		int offSet = itemsLimit;
+		int srcPos = 0;
+
+		if (parent instanceof IExpandableNode expNode) {
+			srcPos = expNode.getOffSet();
+			offSet = srcPos + itemsLimit;
+			// pass on original children
+			sorted = expNode.getAllElements();
+		}
+
+		Object[] partialChildren = new Object[itemsLimit + 1];
+
+		// Extract a subset of children
+		System.arraycopy(sorted, srcPos, partialChildren, 0, itemsLimit);
+
+		// Add an expandable node
+		partialChildren[itemsLimit] = createExpandableNode(sorted, offSet, getItemsLimit());
+
+		return partialChildren;
+	}
+
+	/**
+	 * Concrete viewer is supposed to "expand" {@link IExpandableNode} located at
+	 * given cell.
+	 * <p>
+	 * Default implementation does nothing.
+	 *
+	 * @param cell selected on click
+	 * @since 3.31
+	 */
+	protected void handleExpandableNodeClicked(Widget cell) {
+		// default implementation does nothing. Actual viewers can decide how to
+		// populate remaining elements.
+	}
+
+	/**
+	 * If the items limit is not set, or the number of visible children is below the
+	 * items limit, this method does nothing and returns {@code null}.
+	 * <p>
+	 * In other case, this method tries to fetch (probably updated) model elements
+	 * up to the number of given visible elements, expanding possible existing
+	 * {@link IExpandableNode} elements if needed.
+	 * <p>
+	 * Implementation note: we should not dispose already visible items and we
+	 * should display limited elements along with already visible items.
+	 *
+	 * @param parent          model item to be refreshed.
+	 * @param visibleChildren currently visible children. This includes any elements
+	 *                        already expanded by user.
+	 * @return list of children to be displayed/refreshed or {@code null} if given
+	 *         items length does not exceed items limit
+	 */
+	Object[] getChildrenWithLimitApplied(Object parent, Item[] visibleChildren) {
+		final int limit = getItemsLimit();
+		final int visibleItemsLength = visibleChildren.length;
+		if (visibleItemsLength <= limit || limit <= 0) {
+			return null;
+		}
+		// if we are here, at least one iteration of expandable items is shown
+
+		Object[] currentSorted = getSortedChildren(parent);
+		// if current children count is below limit just return them
+		if (currentSorted.length < limit) {
+			return currentSorted;
+		}
+
+		// numberOfTimesRead is equal to number of times it was expanded by user.
+		int numberOfTimesRead = visibleItemsLength / limit;
+		if ((visibleItemsLength % limit) > 0
+				&& !(visibleChildren[visibleItemsLength - 1].getData() instanceof IExpandableNode)) {
+			numberOfTimesRead++;
+		}
+		List<Object> newChildren = new ArrayList<>();
+		Object lastElement = parent;
+		while (numberOfTimesRead > 0) {
+			Object[] sortedChildren;
+			if (lastElement == parent) {
+				sortedChildren = currentSorted;
+			} else {
+				sortedChildren = getSortedChildren(lastElement);
```

Comment:
This also does too much. We don't need to get sorted children if we already have them in the IExpandableNode

Reply:
Looks like my recent optimization broke the test (sorry, I haven't checked that)
```
java.lang.ArrayIndexOutOfBoundsException: arraycopy: last source index 16 out of bounds for object array[14]
	at java.base/java.lang.System.arraycopy(Native Method)
	at org.eclipse.jface.viewers.ColumnViewer.applyItemsLimit(ColumnViewer.java:899)
	at org.eclipse.jface.viewers.ColumnViewer.getChildrenWithLimitApplied(ColumnViewer.java:966)
	at org.eclipse.jface.viewers.AbstractTableViewer.internalRefreshAll(AbstractTableViewer.java:691)
	at org.eclipse.jface.viewers.AbstractTableViewer.internalRefresh(AbstractTableViewer.java:639)
	at org.eclipse.jface.viewers.StructuredViewer.lambda$3(StructuredViewer.java:1491)
	at org.eclipse.jface.viewers.StructuredViewer.preservingSelection(StructuredViewer.java:1397)
	at org.eclipse.jface.viewers.StructuredViewer.preservingSelection(StructuredViewer.java:1358)
	at org.eclipse.jface.viewers.StructuredViewer.refresh(StructuredViewer.java:1491)
	at org.eclipse.jface.viewers.ColumnViewer.refresh(ColumnViewer.java:555)
	at org.eclipse.jface.viewers.StructuredViewer.refresh(StructuredViewer.java:1452)
	at org.eclipse.jface.viewers.StructuredViewer.setFilters(StructuredViewer.java:1584)
	at org.eclipse.jface.tests.viewers.TableViewerWithLimitTest.testApplyFilter(TableViewerWithLimitTest.java:133)
```